src/chapters/setup-intro-data-types/alternate-code-practice.rst

```diff
@@ -0,0 +1,406 @@
+``fahrenheit`` was declared as a ``double``.
+
+Because Java is a statically typed language, we must call the appropriate
+method on the ``Scanner`` object to ensure the entered data is of the correct
+type. In this case, the compiler compares the types for ``fahrenheit`` and
+``input.nextDouble()`` and throws an error if the two do not match.
+
+The table below shows some commonly used methods of the ``Scanner`` class.
+There are many others supported by this class, and we will talk about how to
+find them in the next chapter.
+
+   TODO: check these descriptions for primitive vs object type usage??
+
+.. list-table:: ``Scanner`` methods
+   :header-rows: 1
+
+   * - Method Name
+     - Return Type
+     - Description
+   * - ``hasNext()``
+     - ``boolean``
+     - Returns ``true`` if more data is present.
+   * - ``hasNextInt()``
+     - ``boolean``
+     - Returns ``true`` if the next item to read is an integer.
+   * - ``hasNextFloat()``
+     - ``boolean``
+     - Returns ``true`` if the next item to read is a ``float``.
+   * - ``hasNextDouble()``
+     - ``boolean``
+     - Returns ``true`` if the next item to read is a ``double``.
+   * - ``nextInt()``
+     - ``Integer``
+     - Returns the next item to read as an ``Integer`` data type.
+   * - ``nextFloat()``
+     - ``Float``
+     - Returns the next item to read as a ``Float`` data type.
+   * - ``nextDouble()``
+     - ``Double``
+     - Returns the next item to read as a ``Double`` data type.
+   * - ``next()``
+     - ``String``
+     - Returns the next item to read as a ``String`` data type.
+   * - ``nextLine()``
+     - ``String``
+     - Returns the next line to read as a ``String`` data type.
+
+Closing ``Scanner``
+^^^^^^^^^^^^^^^^^^^^
+
+To collect data from the command line or other source, create a ``Scanner``
+object. This opens up resources in your machine to manage the input, and these
+resources remain open even after the required data is loaded into your program.
+
+Leaving a ``Scanner`` open is like keeping a window open in your house 24/7.
+Anyone can climb into your home, and you lose $$$ by trying to heat and cool
+your space while it is open to the outside air. Similarly, an open ``Scanner``
+can allow unintended access to your program, and it ties up resources that
+might be needed elsewhere.
+
+Best practice states that if you open a ``Scanner`` object, close it after
+it finishes its job. Line 12 does this in our ``TempConverter`` program:
+
+.. sourcecode:: java
+   :lineno-start: 12
+
+   input.close();
+
+The general syntax is ``scannerObjectName.close()``.
+
+Moving Beyond the Command Line
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The ``Scanner`` class serves as a kind of adapter that gathers primitive data
+types as input and converts them into object types (e.g. it converts an ``int``
+into ``Integer``). We will discuss the purpose of this later, but for now,
+know that this adaptation makes these low level objects easier to use.
```

Comment:
i may have written this part but reading it now, we should maybe change this to read, 
"makes these low-level data types easier to use." to avoid some confusion hopefully

Reply:
Fixed.  Thanks.